ruby-on-rails/ruby-on-rails-quiz.md

```diff
@@ -230,8 +230,8 @@ end
 ```
 
 - [ ] The product will not be updated and the edit template will be rendered.
-- [x] The product will not be updated and the controller will raise an ActiveModel::ForbiddenAttributes exception.
-- [ ] The product will be updated with the values for name, style, and color, but the value for price will be ignored.
+- [ ] The product will not be updated and the controller will raise an ActiveModel::ForbiddenAttributes exception.
+- [x] The product will be updated with the values for name, style, and color, but the value for price will be ignored.
```

Comment:
Before in Rails 4, the B was the answer because it defaulted to always raising an error when unpermitted parameters were attempted to used to create/update a record.

However the default today is to only log, and this can be changed using `config.action_controller.action_on_unpermitted_parameters = :raise` as can be seen [here](https://guides.rubyonrails.org/configuring.html#config-action-controller-action-on-unpermitted-parameters).

Reply:
I feel attacked because the [rails guide](https://guides.rubyonrails.org/getting_started.html#using-strong-parameters) says the exception is thrown, but when I test this, this is not what happens.

Even if I use the shebang save! in my create controller, only this log occurs without any exception being raised, (the data is saved as long as the unpermitted param is not required, a ActiveRecord::RecordInvalid is raised in case I use save! instead of save, but that's not related).

![image](https://github.com/Ebazhanov/linkedin-skill-assessments-quizzes/assets/36988302/8d0d42a6-c151-427b-b427-caf0fc333a47)

Reply:
I decided to make the change from `update_attributes` to `update` as this is better for both old as new versions of Rails. And I decided to correct the answer based on testing (in real life, I can only remember trying to debug why a piece of data is not being saved to database exactly due to this exception not being raised).

I'm down with keeping the original answer as well, given that apart from the test I did, I couldn't come up with some solid source backing me up.

Reply:
I could reproduce what the guides say about raising the exception, but there's a problem. in a controller, that's not the default behaviour used.

![image](https://github.com/Ebazhanov/linkedin-skill-assessments-quizzes/assets/36988302/def23968-329c-4fa5-8ae6-49cc1e608cac)

In the controller, first of all, the params itself is a hash that contains some other data apart from the submitted form.

```ruby
11:04:41 web.1  |   Parameters: {"authenticity_token"=>"[FILTERED]", "card_transaction"=>{"user_id"=>"1", "ct_description"=>"123", "ct_comment"=>"1231", "user_card_id"=>"5", "price"=>"111", "date"=>"2024-02-15"}, "commit"=>"Submit"}
```

As you can see, what we actually get is params[:card_transaction], for example. Inside that, we get the attributes. And that is what `params.require(:card_transaction)` does.

My test in the console was as follows:

![image](https://github.com/Ebazhanov/linkedin-skill-assessments-quizzes/assets/36988302/2c40b5c0-81fb-4209-88db-ac2efb15c7d9)

The reason for this is simple: the `user_params` we use for `User.new` is the result of `params.require(:user).permite(:email)` which **excludes** the entries that are not permitted in the original `params` and therefore, the error is not caught [here](https://github.com/rails/rails/blob/ed76d0fbc74545f0275506b4ebf2dfd375fdda18/activemodel/lib/active_model/forbidden_attributes_protection.rb#L23).

Reply:
I'm very sorry for making this so big. [😔](https://emojipedia.org/pensive-face)